Add more data about the image as block props in the image block

azaozz opened the pull request:
This is the second "chunk" split from #11377. It contains:
1. Adds `fileWidth`, `fileHeight`, `editWidth`, `userSet` block attributes. These would allow us to add width and height, and the themes to generate better `sizes` attributes on the front-end.
    - `fileWidth` and `fileHeight` are the image file dimensions at the time when the image was inserted. For "local" images they can also be "extracted" from the image meta, however there are edge cases where that meta may have changed since the post was published. Also needed for non-resized external images where we don't know the width and height.
     - `editWidth` holds the block width inside the editor at the moment the post was created/edited. Can be used to scale images on the front-end when the block widths don't match (in themes with fluid width or after changing the theme).
      - `userSet` indicated when the user has set the width and height manually. In these cases the theme should (probably) honor the settings when scaling.
2. Fixes the 25, 50, 75, 100% buttons to scale the image relative to the block width. This makes it "visual". Works well in combination with `editWidth` when scaling images on the front-end. Also images "line-up" when setting them to the same scale.
3. Fixes resizing by dragging to prevent upscaling.
4. Fixes direct input resizing to display values only when they were entered by the user.
5. Makes the image block non-dynamic (compared to #11377) and uses the new `render_block` filter to add attributes to the img tag on the front-end, including `srcset` and `sizes`.

azaozz committed:
Add fileWidth, fileHeight, userSet, editWidth props

azaozz committed:
Merge branch 'master' into fix/image-block-add-props

azaozz committed:
Fix upscaling

azaozz committed:
Fix param/prop names

mor10 commented:
I'm trying to test this but am having no luck. Probably doing something stupid or wrong:

```php
function twentynineteen_calculate_image_sizes_attr( $image_attributes, $block_attributes ) {
	var_dump($image_attributes);
	var_dump($block_attributes);
}

add_filter( 'render_block_core_image_tag_attributes', 'twentynineteen_calculate_image_sizes_attr', 10, 2 );
```
This is just a rough test to see what is contained within the attributes so I can build my conditions. Running the above code in WP 4.9.8 and this fork does nothing. Various permutations have not resulted in anything happening when calling the `render_block_core_image_tag_attributes` filter.

Like I said, probably doing something wrong. Please advise.

mor10 commented:
@azaozz What, in your opinion, should be the preferred method for themes to control the `sizes` attribute? The `wp_calculate_image_sizes` filter or the `render_block` filter? Is one more robust / performant / future proof than the other?

azaozz commented:
@mor10 these filters have different scopes, depends what the theme wants to accomplish :)

- The new `render_block_core_image_tag_attributes` gives full control of the `<img>` tag on the front-end (after this PR is merged of course). The theme should use this filter if it needs to set more/other attributes, or to re-calculate some of the existing attributes except `width`, `height`, `srcset`, and `sizes`. These four have their own filters that (will) get the same data, the existing filters for `srcset` and `sizes` plus the new `gutenberg_get_image_width_height` (which would be renamed after merging). Another difference is that `render_block_core_image_tag_attributes` runs only in WP 5.0+ and only when rendering the image block. In the future it may run for other blocks that contain images (Cover, Media and Text), or we may add other filters there..?

- The `wp_calculate_image_sizes` is preferred when the theme looks at the `sizes` attribute. It will get the `$block_attributes` once this PR is merged, which will (hopefully) be enough to generate a much more precise `sizes` attribute. It also exists in older WP versions making it more compatible.

Ideally every theme that supports the Block editor would hook into `gutenberg_get_image_width_height` (as it controls image scaling and generation of `srcset` and `sizes`) and `wp_calculate_image_sizes`. These two would give "full control" over sizing of images on the front-end. Alternatively it can use `render_block_core_image_tag_attributes` and re-calculate some of the img tag attributes (it runs "after the fact" so it will be a tiny bit slower).

Thinking we will have to post something like this explanation on /make as soon as this PR is merged. Would you be up for that? :)

azaozz committed:
Merge branch 'master' into fix/image-block-add-props

azaozz committed:
Better filter name, allow valueless attributes

mor10 commented:
I can help write a post explaining all this. Will test the different approaches first thing next week to get a clear picture of what each of the filters does and how best to guide theme developers to get this baked in before launch.

aduth commented:
> 1. Adds `fileWidth`, `fileHeight`, `editWidth`, `userSet` block attributes. These would allow us to add width and height, and the themes to generate better `sizes` attributes on the front-end.

I have a hard time imagining there's not at least some redundancy in having all of `width`, `fileWidth` (`fileHeight`), `userSet`, and `editWidth`.

- Can `userSet` be derived as a function of the presence of `width` or `height`?
- When would `editWidth` not be equal to `width` in scenarios where e.g. it'd be used for scaling?
- Does it matter what the `fileWidth` and `fileHeight` are for external images? Can that be determined server-side (cached if necessary) if it's relevant there?
   - At worst, could it be reflected in `width` and `height`?

azaozz commented:
Of course there is some redundancy, but it's a "good thing" imho :)

> Can userSet be derived as a function of the presence of width or height?

Sure, except when were width and height set from resizing by dragging, and when were they set directly by the user. In some cases that may not matter in others (scaling on the front end) it's pretty critical.

> When would editWidth not be equal to width in scenarios where e.g. it'd be used for scaling?

Pretty much all resized images? Both "100% width" and resized images can potentially be scaled.

> Does it matter what the fileWidth and fileHeight are for external images?

Yes, since the image block doesn't add width/height for non-resized images. There must be some way for the theme (front-end) to get at least the actual image dimensions. If these are known, the theme (or core) can decide to scale these images too. Also, images outputted without width and height attributes behave "badly", make the page "jump" while loading, push the comments form below the fold (while the user is trying to type), etc.

> Can that be determined server-side (cached if necessary) if it's relevant there? 

Possibly, but to store it we will have to change the image meta schema.

> At worst, could it be reflected in width and height?

Sure. That would mean the image block will always add `width` and `height` attributes to all images? That behavior seems to be the default in pretty much all editors I've seen, not sure why this was removed from the image block... 

However, even when having width and height we would still prefer to know the fileWidth and fileHeight at least for external images. And, if we add them for external, why not for internal? :)

I don't see any problem in saving some extra/redundant image meta in the image block props. This will make in more futureproof. Not having enough meta is much worse (see the image attachments meta where we could have stored few more very very useful values).

mtias commented on packages/block-library/src/image/edit.js:
```diff
@@ -236,27 +284,129 @@ class ImageEdit extends Component {
-		this.props.setAttributes( { url, width: undefined, height: undefined } );
+	setWidthHeight( width, height, fileWidth, fileHeight, userSet ) {
+		// Set image size and also whether set directly by the user.
+		userSet = userSet || undefined;
```
Is this needed?

dmsnell replied:
if this is what I'm thinking it is, it sounds like it's something like `intentionallySetWidth` and `intentionallySetHeight` or `displayWidth` and `displayHeight`

is there confusion since this boolean value doesn't entirely describe why it exists or what the user set?

azaozz replied:
You mean `userSet = userSet || undefined;` or `userSet`? The first is probably not, just some "normalization", but as that's an internal function it can be skipped.

The `userSet` indicates whether the image `width` and `height` were set directly by the user (typed in the width or height fields in the inspector), or they were added as a result of resizing the image by dragging. This would make a difference when the theme or a plugin is deciding to scale or not to scale an image on the front-end.

In the previous iteration these were actually two values: `userWidth` and `userHeight`. Switched it to one boolean after some critique that we store too much image meta in the block props. Can revert that easily :)

aduth replied:
Worth noting there's a subtlety here in that if the argument was called with `userSet = false`, it would normalize the value to `undefined`, which in practice has the same result since the default of the attribute is `false`. And while that default is unlikely to change, **if** it were to, the normalizing would have the adverse effect of preventing a value of explicit `false` to be provided as an argument.

dmsnell commented on lib/compat.php:
```diff
@@ -304,3 +304,235 @@ function gutenberg_warn_classic_about_blocks() {
+	}
+
+	return $content;
+}
```
is there a reason for this not to be hooking into the block filter so that we can get the actual attributes?

```php
function _gutenberg_cache_images_meta( $html, $block ) {
	if ( 'core/image' === $block[ 'blockName' ] ) {
		_prime_post_caches( $block[ 'attrs' ][ 'id' ], /* why aren't these */ false, /* commented? */ true );
	}

	return $html;
}

add_filter( 'rendered_block', '_gutenberg_cache_images_meta', 3 );
```

azaozz replied:
The purpose here is to get the IDs of all images (where we will do srcset and sizes) from the post before we parse the blocks. That's needed to get all image attachment meta in one go from the DB as that's faster. 

So we need an array with all attachment IDs for the images used in the post. This is equivalent to https://core.trac.wordpress.org/browser/tags/4.9.8/src/wp-includes/media.php#L1336.

That should also include images from "Cover" and "Media and Text" blocks, and probably the Gallery block (if images there are "hardcoded"). Any ideas how to do that without a regex are very welcome :)

aduth replied:
Previously: https://github.com/WordPress/gutenberg/pull/11377#discussion_r234008623

I'm not sure I understand why this would need to occur prior to the parse. If it at least occurs prior to _any block evaluating its `render_callback`_, wouldn't that be equally sufficient to warm the cache?

aduth replied:
>If it at least occurs prior to _any block evaluating its `render_callback`_, wouldn't that be equally sufficient to warm the cache?

Since this doesn't exist as an option today anyways (requires filter on [this value](https://github.com/WordPress/WordPress/blob/bf1668333f2da64036db1b10604de18bbbfb8099/wp-includes/blocks.php#L247)), we probably can't do anything of this sort right now.

Should follow-up with a Trac ticket as appropriate.

azaozz replied:
Yeah, this doesn't do anything if we do it "per block". It makes sense only when it happens for all of post_content at once.

dmsnell replied:
if we hook into the block-level filter (the design of which I favor in #10108) then we can track the parse through the blocks and then prime the cache on the `do_blocks` filter which occurs after parsing and rendering is complete.

that is, we could imagine something like this…

```php

function _gutenberg_cache_images_meta( $html, $block ) {
	if ( $block['blockName'] === 'core/image' ) {
		$this->add_image( $block['attrs']['id'] );
	}

	return $html;
}
add_filter( 'rendered_block', '_gutenberg_cache_images_meta', 3 );

function _gutenberg_prime_cache() {
	if ( empty( $this->images ) ) {
		return;
	}

	// whatever command does all images in one go
	_prime_post_caches( $this->images, false, true, true, true, false, false );
}
add_filter( 'do_blocks', '_gutenberg_prime_cache' );
```

and this because we know that the final `do_blocks` filter runs _after_ block-level processing.

dmsnell replied:
noting that this is still based on a RegExp instead of using the per-block and per-document filters

azaozz replied:
Another option would be to "unify" this with the "old" cache priming. We have the (required) `wp-image-####` className on all images, including in Cover and Media and Text blocks.

Will need to abstract [`wp_make_content_images_responsive()`](https://core.trac.wordpress.org/browser/tags/4.9.8/src/wp-includes/media.php#L1305) a bit, move the caching out and run it earlier on `the_content` filter.

On the other hand `_prime_post_caches()` is "clever enough" to not fetch posts and meta that are already cached so running two pre-caching methods won't "hurt" anything.

dmsnell commented on packages/block-library/src/image/edit.js:
```diff
@@ -236,27 +284,129 @@ class ImageEdit extends Component {
 
-	updateImageURL( url ) {
-		this.props.setAttributes( { url, width: undefined, height: undefined } );
+	setWidthHeight( width, height, fileWidth, fileHeight, userSet ) {
```
this name is definitely confusing here…does the width have a height?

dmsnell commented on packages/block-library/src/image/edit.js:
```diff
@@ -236,27 +284,129 @@ class ImageEdit extends Component {
+			width: undefined,
+			height: undefined,
+			userSet: undefined,
+		} );
```
there's no need here for two calls to `setAttributes()` - we can collapse them into one in a couple of different ways. my favorite is with `Object.assign()`

```js
// what does this name mean?
resetWidthHeight( fileWidth, fileHeight ) {
	this.props.setAttributes(
		{
			width: undefined,
			height: undefined,
			userSet: undefined,
		},
		fileWidth && fileHeight && {
			fileWidth,
			fileHeight,
		}
	);
}
```

azaozz replied:
Sure. The purpose here is to make it as easy to read and understand as possible. After going through babel and minification, the code that actually runs in production will be completely different.

(I actually "have a beef" with the way a lot of the code is written. All the redundant indentation, nested lambda functions (no names = poor context), "clever" conditionals, almost no inline comments... There seem to be quite a few cases where the code becomes more readable after running it through bable and minification, then "unminifying" it. But lets look at that after 5.0 is out.) :)

aduth replied:
> there's no need here for two calls to `setAttributes()` - we can collapse them into one in a couple of different ways. my favorite is with `Object.assign()`

(Noting that the supplied example doesn't use `Object.assign` and wouldn't work as written)

dmsnell replied:
> (Noting that the supplied example doesn't use Object.assign and wouldn't work as written)

Good call! 😆

aduth replied:
> Sure. The purpose here is to make it as easy to read and understand as possible. After going through babel and minification, the code that actually runs in production will be completely different.

The main thing for me here would be to avoid multiple calls to `setAttributes`, as it will incur a render for each (unnecessary impact on performance). This will remain true after transpilation and minification.

I pushed bf67408 which I think is a reasonable compromise to preserve a similar readability.

dmsnell commented on packages/block-library/src/image/edit.js:
```diff
@@ -236,27 +284,129 @@ class ImageEdit extends Component {
+					fileWidth = imageData.width;
+					fileHeight = imageData.height;
+					return false;
+				}
```
using `map` without returning a new list and when modifying things inline is pretty unusual. we have a couple of choices depending on what our intention is. it looks like we're looking for a specific value and using that if it exists.

```js
updateImageURL( url, sizeOptions ) {
	const { width, height } = sizeOptions;

	if ( width && height ) {
		return this.setAttributes( {
			url,
			fileWidth: width,
			fileHeight: height,
		} );
	}

	const option = find( sizeOptions, ( { value } ) => value === url );

	if ( option ) {
		return this.setAttributes( {
			url,
			fileWidth: option.imageData.width,
			fileHeight: option.imageData.height;
		} );
	}
}
```

dmsnell commented on packages/block-library/src/image/edit.js:
```diff
@@ -170,10 +195,30 @@ class ImageEdit extends Component {
+			img = media.sizes.large || media.sizes.full;
+			src = img.url;
+			actualWidth = get( img, [ 'actual_size', 'width' ] ) || img.width;
+			actualHeight = get( img, [ 'actual_size', 'height' ] ) || img.height;
```
`get()` has a built-in default parameter for the case where the value doesn't exist or is `undefined`. I'm not sure if we'd expect to get `actual_size.width === 0` here but if we don't then we can use the default param.

```js
get( img, 'actual_size.width', img.width );
get( img, 'actual_size.height', img.height );
```

azaozz replied:
Actually seems we will need to take that one step further. Need to ensure both `get( img, [ 'actual_size', 'width' ] )` and `get( img, [ 'actual_size', 'height' ] )` are present before using them, or fall back to `img.width` and `img.height`. The image meta in the DB can be really inconsistent in some rare cases like old plugins mangled it, etc.

aduth replied:
> Actually seems we will need to take that one step further. Need to ensure both `get( img, [ 'actual_size', 'width' ] )` and `get( img, [ 'actual_size', 'height' ] )` are present before using them, or fall back to `img.width` and `img.height`. The image meta in the DB can be really inconsistent in some rare cases like (old) plugins mangled it, etc.

Condition enforced as of 4d2d959.

dmsnell commented on packages/block-library/src/image/edit.js:
```diff
@@ -170,10 +195,30 @@ class ImageEdit extends Component {
+			// Not used in the editor, passed to the front-end in block attributes.
+			fileWidth: actualWidth,
+			fileHeight: actualHeight,
+			editWidth: blockWidth,
```
I'm finding some of the new mutation difficult to follow - to see what it's supposed to be setting. I _think_ we're wanting to set the actual image size if given, else use the default sizes as if none is given?

```js
const { sizes: givenSizes } = media;
const givenSize = givenSizes && ( givenSizes.large || givenSizes.full );

this.props.setAttributes( Object.assign(
	{
		url: media.url,
		editWidth: getBlockWidth(),
	},
	pickRelevantMediaFiles( media ),
	givenSizes && {
		url: size.url,
		fileWidth: get( givenSize, 'actual_size.width', givenSize.width ),
		fileHeight: get( givenSize, 'actual_size.height', givenSize.height ),
	}
} );
```

azaozz replied:
The whole problem is that the image meta in the media modal is "tinted". The sizes are passed through [image_constrain_size_for_editor()](https://core.trac.wordpress.org/browser/tags/4.9.8/src/wp-includes/media.php#L27) which uses the global `$content_width` to limit them. However we want the "normal" sizes, so had to add https://github.com/WordPress/gutenberg/pull/11973/files/03444d7bdda0d79dcc540d0fbfa5141c2a14de78#diff-6ff32417da0658502e7caa1a1abbeae6R309 to fix it.

It's not really a mutation, it tries to pick one value if it exists, and falls back to another no matter what it is. Can be written in several ways, the wordier one seems more readable :)

dmsnell commented on packages/block-library/src/image/edit.js:
```diff
@@ -145,6 +151,25 @@ class ImageEdit extends Component {
+					// Add image file dimensions.
+					this.props.setAttributes( {
+						fileWidth: get( sizeFull, [ 'actual_size', 'width' ] ) || sizeFull.width,
+						fileHeight: get( sizeFull, [ 'actual_size', 'height' ] ) || sizeFull.height,
```
same note here as in the one below about `get()` and its default value…

```js
get( sizeFull, 'actual_size.width', sizeFull.width )
```

azaozz replied:
Yes, preferably to test for "falsy" value from `get()`.

dmsnell commented on packages/block-library/src/image/edit.js:
```diff
@@ -8,6 +8,7 @@ import {
 	map,
 	last,
 	pick,
+	round,
```
why did we bring in `lodash.round` instead of using `Math.round()` the native function?

azaozz replied:
Uh, that was used in previous iterations to round to specific precision. Not used any more so can remove it :)

aduth replied:
> Uh, that was used in previous iterations to round to specific precision. Not used any more so can remove it :)

Changed to `Math.round` in 32d5c8c

dmsnell commented on lib/compat.php:
```diff
@@ -304,3 +304,235 @@ function gutenberg_warn_classic_about_blocks() {
+	$attr = '';
+	foreach ( $image_attributes as $name => $value ) {
+		// Sanitize for valid HTML 5.0 attribute names, see: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes.
+		$name = preg_replace( '/[^a-z0-9-]+/', '', strtolower( $name ) );
```
should we just remove it if it's not valid? seems like this could introduce an XSS error?

```html
<img on&click="alert('hi');" />
```

by removing the unwanted letters we might create attributes we shouldn't. can you think of any valid reason to pass through an attribute that's invalid?

azaozz replied:
Hmm, good point. Will change to remove when invalid chars. The only exception (well, more of a limitation) may be for `data-*` attribute names, but thinking we can limit these to `a-zA-Z0-9-` too.

aduth replied:
Out of curiosity, what's the impact of letting this go through unsanitized. These are generated by plugins?

If not for WordPress' lack of sanitization in other filtered functions, I'd not be able to do fun hacks like:

https://github.com/aduth/dones/blob/0450c88429deb4546f72d20f915551ec62a33de3/functions.php#L215

aduth replied:
There's some prior art to work from here in `@wordpress/escape-html` :

https://github.com/WordPress/gutenberg/blob/563ef41c663f163a725c2ef7e961b11ea6981f6e/packages/escape-html/src/index.js#L66

Is there not some existing PHP function we could be using?

azaozz replied:
> what's the impact of letting this go through unsanitized

Right, this comes from plugins, but core (often) sanitizes values in these cases. 

> s there not some existing PHP function

There are couple but not exactly the same. We should probably introduce one. As @dmsnell points out it should fail the test when unexpected chars, not strip them.

aduth replied:
> As @dmsnell points out it should fail the test when unexpected chars, not strip them.

Attributes stripped in 8e73feb.

aduth replied:
> > The only exception (well, more of a limitation) may be for data-* attribute names
> 
> If we want to conform to the HTML spec we can do just that - it's not hard

Same RegExp exists in `escape-html` as well

https://github.com/WordPress/gutenberg/blob/563ef41c663f163a725c2ef7e961b11ea6981f6e/packages/escape-html/src/index.js#L12

aduth replied:
> > As @dmsnell points out it should fail the test when unexpected chars, not strip them.
> 
> Attributes stripped in [8e73feb](https://github.com/WordPress/gutenberg/commit/8e73feb27668b3837aae78ba6d59fab565d734c1).

Oops, argument order was incorrect. I'll account for this in a follow-up commit which includes use of the above regular expression.

aduth replied:
> > > As @dmsnell points out it should fail the test when unexpected chars, not strip them.
> > 
> > 
> > Attributes stripped in [8e73feb](https://github.com/WordPress/gutenberg/commit/8e73feb27668b3837aae78ba6d59fab565d734c1).
> 
> Oops, argument order was incorrect. I'll account for this in a follow-up commit which includes use of the above regular expression.

Corrected (with new pattern) in 19dad07.

azaozz replied:
Hm, this will "let through" a lot chars that are not present in any HTML 5.0 attribute :) 
The intention here is to "Sanitize for valid HTML 5.0 attribute names".

All attributes there (except `data-*`) use only lower case `a-z`, `0-9` and `-`. If we really want to allow `data-*` to contain all allowed chars in attribute names (this is not the same as "valid attribute names"), then we would need to make an exception only for it.

Also `$is_invalid_attribute` should probably be `$is_invalid_attribute_name`, we only check the names there.

Frankly I'd prefer to "enforce" use of only `a-z0-9-` even for the variable part of `data-*`. This is already done in other places in core, don't see a compelling reason for it to be different here.

azaozz replied:
Updated in https://github.com/WordPress/gutenberg/pull/11973/commits/e3a99c74079627f9037551efe30149b7f5665fc0 to check for `[^a-z0-9-]` with an exception for `data-*`. Also fixed the replace image regex a bit, we want at least one white space char after `<img`.

azaozz commented on packages/block-library/src/image/edit.js:
```diff
@@ -236,27 +284,129 @@ class ImageEdit extends Component {
+
+		return {
+			width: width,
+			height: height,
```
`constrainImageDimensions` is pretty much a copy of core's `wp_constrain_dimensions()`: https://core.trac.wordpress.org/browser/tags/4.9.8/src/wp-includes/media.php#L400. We can change/improve it as long as it works in exactly the same way (and is "easy to read") :)

> since this doesn't depend on any instance variables, might as well put it outside of the class and make it a simple helper function

Sure, it probably can be re-used somewhere else.

aduth committed:
Merge branch 'master' into fix/image-block-add-props

aduth committed:
Compat: Resolve PHP lint errors

aduth commented on lib/compat.php:
```diff
@@ -304,6 +304,247 @@ function gutenberg_warn_classic_about_blocks() {
+		( 'wide' === $block_attributes['align'] || 'full' === $block_attributes['align'] ) &&
+		! empty( $block_attributes['fileWidth'] ) &&
+		$block_attributes['fileWidth'] < $image_meta['width'] &&
+		max( (int) $image_meta['width'], (int) $image_meta['height'] ) < 4300 && // max 12 MP photo (that's still may be over 3MB, consider reducing).
```
This line gives me some pause because it introduces caveats to the claim that "Full wide alignment uses the original image", the number itself maybe (inherently) arbitrary, and very much a magic number (should be constant? filtered value?).

azaozz replied:
Yeah, that's a pretty... arbitrary assumption. We just need a "sensible default" for now. 

Really hoping we can add another larger image size that's generated by default in core, like `xlarge`. Then we can just use it or fall back to "full" when it doesn't exist.

azaozz replied:
Updated this to look for larger image sizes that were added by themes or plugins first, see https://github.com/WordPress/gutenberg/pull/11973/commits/44c7c68ae058792d65e88553f526df44358d5b32.

aduth commented on lib/compat.php:
```diff
@@ -304,6 +304,247 @@ function gutenberg_warn_classic_about_blocks() {
+		$block_attributes['fileWidth'] < $image_meta['width'] &&
+		max( (int) $image_meta['width'], (int) $image_meta['height'] ) < 4300 && // max 12 MP photo (that's still may be over 3MB, consider reducing).
+		! empty( $image_meta['sizes']['large'] ) &&
+		$block_attributes['fileWidth'] === (int) $image_meta['sizes']['large']['width'] &&
```
I'm not clear why this condition exists if at least the next line assures an equal aspect ratio.

azaozz replied:
It's an edge case check. Image attachment meta in core may be... inconsistent in rare cases. This is an "just in time" check to make sure we don't trigger PHP warnings.

aduth committed:
Compat: Correct closing img tag

aduth commented on lib/compat.php:
```diff
@@ -304,6 +304,247 @@ function gutenberg_warn_classic_about_blocks() {
+	$image_tag = '<img' . $attr . '/ >';
+
+	// Replace the img tag.
+	$html = preg_replace( '/<img [^>]+>/', $image_tag, $html );
```
Should we require the space after `img` ? Unlikely it'd ever occur, just caught my attention considering we just finished constructing a string which could potentially be `<img/>`.

azaozz replied:
"Best practice" would be to require it. In this case it doesn't matter but some tags match the beginning of the names of other, like `<b` and `<br`.

It can't be `<img/>` as at least the src should be there :)

aduth replied:
> In this case it doesn't matter but some tags match the beginning of the names of other, like `<b` and `<br`.

The revised implementation should account for this in testing as `\s*`. (04706dd)

azaozz replied:
Should be `\s` (at least one white space char). Changed in https://github.com/WordPress/gutenberg/pull/11973/commits/e3a99c74079627f9037551efe30149b7f5665fc0.

aduth commented on lib/compat.php:
```diff
@@ -304,6 +304,247 @@ function gutenberg_warn_classic_about_blocks() {
+		}
+	}
+
+	$image_tag = '<img' . $attr . '/ >';
```
Noting that this seems like it could serve as a generally reusable tag builder, not too much unlike [`wp.element.renderToString`](https://github.com/WordPress/gutenberg/blob/f91e66ca2bcd9e1e7b97aa39c01f2888fc751712/packages/element/src/serialize.js#L425)

aduth committed:
Compat: Allow optional space after img tag name

aduth committed:
Compat: PHPDoc formatting and arrangement style

aduth commented on packages/block-library/src/image/edit.js:
```diff
@@ -141,7 +142,12 @@ class ImageEdit extends Component {
+		const blockWidth = getBlockWidth();
+
+		// Store the current block width inside the editor.
+		this.props.setAttributes( { editWidth: blockWidth } );
```
A worry here is that if the value wasn't already assigned, it would immediately make the post "dirty" (prompting on navigation, triggering delayed autosave).

aduth replied:
Ideally we shouldn't call `setAttributes` in `componentDidUpdate` because it will trigger a second (ideally unnecessary render):

`[Something Triggers Update] -> [render] -> [componentDidUpdate] -> [setAttributes] -> [render] -> [componentDidUpdate]`

aduth replied:
Furthermore, there's nothing which would assign the `editWidth` attribute when the component first mounts?

azaozz replied:
Hm, is there a better place/way to do this? We need it done once per inserted/uploaded image.

aduth committed:
Blocks: Use Math.round for image rounding

aduth commented on packages/block-library/src/image/index.js:
```diff
@@ -30,15 +30,9 @@ export const name = 'core/image';
 		type: 'string',
-		source: 'attribute',
-		selector: 'img',
-		attribute: 'src',
```
This change will invalidate all existing image blocks.

![image](https://user-images.githubusercontent.com/1779930/48784390-97cd1800-ecb0-11e8-9bcd-6b00f363067b.png)

```
Block validation: Block validation failed for `core/image` ({name: "core/image", title: "Image", description: "Insert an image to make a visual statement.", icon: {…}, category: "common", …}).

Expected:

<figure class="wp-block-image alignwide"><img alt="" class="wp-image-1962"/></figure>

Actual:

<figure class="wp-block-image alignwide"><img src="http://editor.test/wp-content/uploads/2018/11/desk-1024x614.jpeg" alt="" class="wp-image-1962" /></figure>
```

https://wordpress.org/gutenberg/handbook/block-api/deprecated-blocks/

azaozz replied:
Yeah, need to add perhaps `oldBlockAttributes` for use in the deprecated save().

aduth committed:
Blocks: Add deprecation for attribute-sourced URL, alt

aduth commented on packages/block-library/src/image/image-size.js:
```diff
@@ -9,6 +9,11 @@ import { noop } from 'lodash';
 import { withGlobalEvents } from '@wordpress/compose';
 import { Component } from '@wordpress/element';
 
+export function getBlockWidth() {
```
I've confirmed that depending on whether the image block is Wide/Full aligned and whether it's the first block will return a different value. For example, if I've a wide image block as the sole content, this returns `718`. If I insert a paragraph before the image block, it suddenly starts returning `580` (the width of the non-wide paragraph).

We should assign a [`ref`](https://reactjs.org/docs/refs-and-the-dom.html) to the image from which to inspect the DOM width.

aduth replied:
> We should assign a [`ref`](https://reactjs.org/docs/refs-and-the-dom.html) to the image from which to inspect the DOM width.

Except: Is it really the image width we care about? Or is the end-goal to get what is effectively the rendered width of the block list (non-wide, non-full)? (Which I assume would ideally be the same as `$content_width`)

azaozz replied:
> Noting that this would return the width value of any block, not necessarily the image block.

Right, this is the intent. See the comment above.

> a block shouldn't assume anything about its ancestry, as it makes for a maintainability nightmare should we ever change the class of the block wrapper.

> Except: Is it really the image width we care about? Or is the end-goal to get what is effectively the rendered width of the block list (non-wide, non-full)? (Which I assume would ideally be the same as $content_width)

The latter. Unfortunately there's no "easy way" to get the editor "visual" width at any given moment. We need to look at the DOM somewhere. Any suggestions on how to improve that are very welcome :)

The old `$content_width`, even if we reuse it again would not work here as themes can set the width from CSS. My idea was to "fix" the width to content_width or a similar new global that would be used as "max-width" by themes and enforced in the editor and the front-end.

It's probably not too late to do that, just need to set it as expectation for the themes.

aduth replied:
If we needed to reach into the DOM, and it's expected a theme would provide their custom width by overriding the `max-width` style of `.wp-block`, and we have no guarantee there'd be a non-full/wide align block present in the editor, I might imagine there could be an option with creating a DOM node for the sole purpose of inspecting its computed style:

```js
const div = document.createElement( 'div' );
div.className = 'wp-block'
const { maxWidth } = window.getComputedStyle( div );
// '610px'
document.body.removeChild( div );
```

There's a couple problems here:

- It's tragically non-performant if subject to repeated calls
- To get from 610 to 580, we need to subtract the paddings of the block, which vary depending on screen size, and are styled contextualized to `.editor-block-list__layout .editor-block-list__block`

aduth replied:
> If we needed to reach into the DOM, and it's expected a theme would provide their custom width by overriding the `max-width` style of `.wp-block`, and we have no guarantee there'd be a non-full/wide align block present in the editor, I might imagine there could be an option with creating a DOM node for the sole purpose of inspecting its computed style:

I went with something like this in 87130e4, considering it a temporary solution, cached as much as is reasonable (cleared on window resize).

aduth commented on packages/block-library/src/image/edit.js:
```diff
@@ -470,45 +623,57 @@ class ImageEdit extends Component {
-
-										const isCurrent = width === scaledWidth && height === scaledHeight;
+									{ [ 25, 50, 75, 100 ].map( ( percent ) => {
+										const blockWidth = getBlockWidth();
```
There's no way the block could know its own width on an initial render.

azaozz replied:
`getBlockWidth()` doesn't necessarily get this specific block's width. It rather gets the "visual" editor width for all blocks. This is needed as Gutenberg allows themes to set the editor width from CSS, and some (TwentyNineteen) do "fluid" width.

aduth committed:
Blocks: Use hard-coded content width for block width

aduth commented:
The changes in e4cfea3 make an (unconfirmed) assumption that the value of `editWidth` should align to the default block list width (for non-wide, non-full blocks), which itself should _ideally_ (but in practice, not yet) align to the effective value of the `$content_width` global. As written, it's intended to be forward-compatible to a scenario where this is provided through editor settings.

aduth committed:
Blocks: Limit editWith assignment to insert, resize

aduth commented:
> Another question of `editWidth`: At what intervals should it be updated? Block insertion? Resize? An issue with the current implementation is that if the actual value of the width changes in the future (wider / thinner block list area), the block automatically updates itself immediately when the editor loads. This doesn't seem desirable.

As of 4c49a8d, it is assigned on insert and resize.

aduth commented on packages/block-library/src/image/index.js:
```diff
@@ -76,6 +70,19 @@ const blockAttributes = {
+	fileHeight: {
+		type: 'number',
+	},
+	userSet: {
```
While in the context of the pull request the name doesn't draw much attention, but without that as context, it'd be hard to associate this being related to dimensions of the image. I'm thinking maybe a suffix of "Dimensions" ?

azaozz replied:
Was trying to come up with the shortest names that make sense as these will be repeating a lot in all posts. Feel free to change as you like :)

aduth committed:
Blocks: Rename Image userSet to userSetDimensions

aduth committed:
Blocks: Remove redundant userSetDimensions normalization

aduth committed:
Blocks: Document constrainImageDimensions as ported from PHP

aduth committed:
Blocks: Avoid mutative, unused Array#map result

aduth commented on packages/block-library/src/image/edit.js:
```diff
@@ -278,27 +321,139 @@ class ImageEdit extends Component {
+		this.props.setAttributes( { align: nextAlign } );
+	}
+
+	updateImageURL( url, imageSizeOptions ) {
```
What is the expected shape of the second argument here? It seems like it could one or either of an object or an array of said objects? The overloading is hard to follow. For the array form, it also seems like the function could just call to `this.getImageSizeOptions` directly, which would leave open the option for the second argument to serve the sole purpose of being an explicit width/height.

azaozz replied:
> It seems like it could one or either of an object or an array of said objects

Yes, it reuses the array passed to the image sizes drop-down. That was the "original" use. Added support for the object with width and height so it can be reused when updating the URL of uploaded image. Of course can have another function there, or just add that inline.

dmsnell replied:
I'd love to see a second function instead of overloading parameter types. it took me considerable time to try and figure out what was happening here, especially since this PR only calls it with one of the types.

aduth committed:
Block: Avoid overloading Image block updateImageURL

aduth commented on packages/block-library/src/image/edit.js:
```diff
@@ -588,21 +764,35 @@ class ImageEdit extends Component {
 								/* eslint-enable jsx-a11y/no-noninteractive-element-interactions */
 							);
 
+							// Floating a resized image can produce inaccurate `imageWidthWithinContainer`.
```
I'm not really sure what this comment has to do with the logic which follows it. If `imageWidthWithinContainer` is producing an inaccurate value, should that be corrected from within `ImageSize` ? I don't really see how it has any impact one way or the other on how it's used here though.

azaozz replied:
Uh, this comment seems left over from previous iterations... At some point it was using the imageWidthWithinContainer.

aduth committed:
Blocks: Fix Image accidental inline tab character

aduth commented on lib/compat.php:
```diff
@@ -304,6 +304,253 @@ function gutenberg_warn_classic_about_blocks() {
+function _gutenberg_cache_images_meta( $content ) {
+	// Need to find all image blocks and get the attachment IDs from them BEFORE the parser is run
+	// so we can get the image attachments meta all at once from the DB.
+	if ( preg_match_all( '/^<!-- wp:image {.*"id":(\d+),.*} -->$/m', $content, $matches ) ) {
```
Two issues with the regular expression:

- Lazy `.*` will potentially capture across multiple image blocks in the same line
- There's an enforced comma after `id` which assumes that `id` can't be the last property of the JSON (which we shouldn't assume)

https://regex101.com/r/SX8DA0/1

azaozz replied:
> Greedy .* will potentially capture across multiple image blocks in the same line

Yeah, this assumes that the "block boundaries" are formatted properly, i.e. each is on a new line. Then there will only be one image block with one ID. That makes it somewhat fragile, but can't think of anything better that's fast enough... Still, planning to come back to this later (if possible).

aduth committed:
Compat: Avoid assumption of ID as non-last attribute of image

aduth committed:
Compat: Document image block RegEx as temporary solution

aduth committed:
Blocks: Pick url, alt in raw transform from image node

Since these are no longer sourced from attributes and instead assumed to exist in comment attributes, they must be passed explicitly

aduth committed:
Framework: Regenerate fixtures

aduth commented:
Remaining changes are fixtures updates, which are complicated (blocked by?) local changes upon `npm run fixtures:regenerate`. A review of #11973 would avoid the need to include those changes on this branch.

aduth committed:
Blocks: Update image fixtures per url, alt as comment attributes

aduth committed:
Compat: Align PHPDoc return

aduth committed:
Compat: Add PHPDoc since tag

dmsnell commented on packages/block-library/src/image/edit.js:
```diff
@@ -177,10 +198,29 @@ class ImageEdit extends Component {
+			fileHeight = get( img, [ 'actual_size', 'height' ] ) || img.height;
+		}
+
+		const attr = pickRelevantMediaFiles( media );
```
why are we pulling this out? it's still confusing me and makes me think there's something else going on even though to the best of my knowledge there's nothing happening. we can still set the `url` attributes in the `setAttributes()` call just by adding it after `...pickRelevantMediaFiles( media )` and then it won't be splitting the computation or providing a point where we could introduce a data consistency bug

azaozz replied:
Hm, this is actually not needed any more as now `pickRelevantMediaFiles()` properly gets the "large" size of the image if it exists. That was the same "fix".

aduth committed:
Merge branch 'master' into fix/image-block-add-props

aduth committed:
Blocks: Remove outdated comment about floated images width accuracy

aduth committed:
Compat: Remove any attributes containing invalid characters

aduth committed:
Blocks: Reintegrate media attributes pick to setAttributes spread

https: //github.com/WordPress/gutenberg/pull/11973#discussion_r235115253
Co-Authored-By: Dennis Snell <dennis.snell@automattic.com>

aduth committed:
Blocks: Set Image fileWidth, fileHeight only if both present

Co-Authored-By: Andrew Ozz <azaozz@users.noreply.github.com>

dmsnell commented on lib/compat.php:
```diff
@@ -304,6 +304,261 @@ function gutenberg_warn_classic_about_blocks() {
+	$attr = '';
+	foreach ( $image_attributes as $name => $value ) {
+		// Sanitize for valid HTML 5.0 attribute names, see: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes.
+		$is_invalid_attribute = preg_match( '/[^a-z0-9-]+/', '', strtolower( $name ) );
```
(copying here because it was hidden through a code update)

> The only exception (well, more of a limitation) may be for data-* attribute names

If we want to conform to the HTML spec we can do just that - it's not hard

> Attributes have a name and a value. Attribute names must consist of one or more characters other than the space characters, U+0000 NULL, U+0022 QUOTATION MARK ("), U+0027 APOSTROPHE ('), U+003E GREATER-THAN SIGN (>), U+002F SOLIDUS (/), and U+003D EQUALS SIGN (=) characters, the control characters, and any characters that are not defined by Unicode. In the HTML syntax, attribute names, even those for foreign elements, may be written with any mix of lower- and uppercase letters that are an ASCII case-insensitive match for the attribute’s name.
> https://www.w3.org/TR/html52/syntax.html#attribute-names

```js
/[^\u0000-\u0020\u007f-\u009f"'>/=]+/
```

that should define them. that leaves open some valid names that WordPress will sanitize, but they are valid.

azaozz replied:
The intention there was to check for valid HTML 5.0 attribute names only, not all possible attribute names. In HTML 5.0 the only attribute name that may contain any of the valid chars is the `data-*` (in the variable part).

However I think it's better to keep that variable part limited to only `a-z0-9-` like it is at other places in core.

As discussed previously, it would be good to make that into another "filtering" function in core. Perhaps then we can decide whether to support all valid chars or to keep the restriction.

aduth committed:
Compat: Use spec standard for valid attribute characters

aduth committed:
Editor: Compute expected block width by DOM compute

aduth commented on lib/compat.php:
```diff
@@ -304,6 +304,263 @@ function gutenberg_warn_classic_about_blocks() {
+			$value = esc_attr( $value );
+		}
+
+		if ( empty( $value ) ) {
```
Boolean attributes are of a whitelisted set, and I don't think it's safe to treat all empty values in the same sense.

For example, this has the impact of writing `img` tags with `alt` omitted as:

![image](https://user-images.githubusercontent.com/1779930/48800204-1dfc5500-ecd7-11e8-925d-5948a7b22655.png)

https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attributes

https://github.com/WordPress/gutenberg/blob/5a8c012de43780f7187eecf7ac796e6f57a04c4f/packages/element/src/serialize.js#L92-L136

aduth replied:
I think it's probably best to avoid this part of the condition and always write the attribute, even if empty. Particularly since it's safe (semantically equivalent) for boolean attributes to "needlessly" have a value assigned.

_Semantically equivalent:_

```html
<input disabled="disabled">
<input disabled="">
<input disabled>
<input disabled="wut">
```

aduth committed:
Compat: Always write img attribute value, even empty

aduth committed:
Blocks: Avoid multiple calls to setAttributes in resetWidthHeight

aduth commented:
Code-wise I think this is as far as it's going to make sense to get along here. It could do for a fresh shake of thorough testing.

Tasks which should be followed up on:

- Use filter on parsed result of blocks in place of regular expression
- Find a more durable process for measuring the block editor width
- Automated tests

jorgefilipecosta reviewed (commented):
While testing I noticed some possible improvements, they don't seem blockers.
Updating the image using the resizer, and using the image dimensions fields, changes the same attributes. But if we set a value on the fields and then we use the resizer the value that appears on the fields is reset, I think we should show the new width and height set using the resizer.
Depending on the theme the 75% and/or 50% resizing buttons may not work and 100% is used instead.

aduth commented:
Related to @jorgefilipecosta 's comment, from the original message:

>Fixes the 25, 50, 75, 100% buttons to scale the image relative to the block width. This makes it "visual". Works well in combination with editWidth when scaling images on the front-end. Also images "line-up" when setting them to the same scale.

What should happen if the user selects a percentage which is larger than the actual width of the image?

For example, considering a block width of 580px and a "Medium" size image naturally 300px wide, if we select 50%, it works well enough to show at 290px, but then selecting 75% has it move up to only 300px and shows 100% button as selected instead. Technically this aligns with some expectation that the image not be upscaled to exceed its own natural width, but it "feels buggy".

youknowriad commented:
@aduth I guess that's the intentation but I'd argue that we should use the image width as 100% in that case to avoid the "buginess" feeling.

catehstn commented:
@aduth are you happy to merge this now?

aduth commented:
If the change in behavior of applying a percentage width from being a factor of the image's natural width to a factor of the block's width (and never upscaling) is agreeable, then this could probably be merged without additional changes. Additional testing would be good to have.

In retrospect, the specific change feels like something which should probably have been addressed separately, since it doesn't seem directly related to the need to optimize the server-rendered output for the front-end display of the image.

Before end of day yesterday I had caught a glance at the prior core implementation for images, and noted that there's some overlap and some differences. This seems like something which should be aligned; either removing our sanitization here prior to merge, or following-up to do the same in core. I'm not sure why they would need to be different?

https://github.com/WordPress/wordpress-develop/blob/5d92e0c0a950399c53294a41000305ddcf47efb5/src/wp-includes/media.php#L911-L927

There's probably some smaller tasks which could be pulled out of this one, but it may just as well be fine to move forward with all of it as one unit. Other options could include most of the bullet points of the original message, and maybe some more direct (incomplete) approach in adding size as a comment attribute and using that to perform string replacements on the server-rendered result, rather than reconstructing the entire image.

I should note that I am out for holiday through the rest of this week.

azaozz commented:
>For example, considering a block width of 580px and a "Medium" size image naturally 300px wide, if we select 50%, it works well enough to show at 290px, but then selecting 75% has it move up to only 300px and shows 100% button as selected instead. Technically this aligns with some expectation that the image not be upscaled to exceed its own natural width, but it "feels buggy".

> If the change in behavior of applying a percentage width from being a factor of the image's natural width to a factor of the block's width (and never upscaling) is agreeable, then this could probably be merged without additional changes. Additional testing would be good to have.

Went through several iterations, there are downsides to all possible behaviors. The current one is the "most logical" but feels somewhat buggy for images that are narrower than the editor width.

If we go with "double standard" based on whether the image width exceeds the editor width, we end up with inconsistent sizes on the front-end (when images get scaled or for themes with "fluid" width). We also loose the (once very requested) feature to be able to line-up images easily/automatically.

To improve the "feel" for narrower images seems best would be to disable the buttons that don't apply to them, i.e. for 600px wide editor and 500px wide image the 100% button will be disabled. For 200px image the 50, 75, and 100% buttons will be disabled, etc.

youknowriad commented:
My opinion here is that we should seek for clarity which means Add an explicit toggle to the UI saying: "Relative size / absolute size" and all the resizing that can happen in the UI (scale buttons, inputs or the resizer) should do the exact same thing by respecting the user choice in that toggle.

We could default to "relative size" if that's the most desired option.

azaozz commented:
> Add an explicit toggle to the UI saying: "Relative size / absolute size"

Was thinking about something similar too. But there are a lot of downsides to "absolute" resizing.  It would make things very complicated/hard to understand (not so much in the code, but in the UI). Also resizing by dragging is always relative, don't even think we can make it "absolute". We probably can force it in the resizer, but how is that going to work in the UI? It would mean resizing by dragging won't change the visible size of the image until the user reaches the current block width? That would be very weird :)

Another big downside is the user expectation when resizing a large (for example 6000px wide) image by absolute size to 50%? They will get an img tag for a 3000px soft-resized image. That image will be "shrunk" by CSS `max-width` in the editor and (most likely) in the theme on the front-end. The result is: no change in how the image is displayed. (It will also produce a `sizes` attribute of 3000px which will "lock" the image source the browser can choose from the `srcset` to the full size image as the other images are quite smaller.)

So setting a standard photo to "absolute" 50% width will make no visible difference anywhere, and force the browser to always load the (very large, non-optimized) original image. Don't really see any user cases for such behavior. The opposite is true, we should try to prevent this from happening :)

Even if the user has edited properly sized an image before uploading it, and still want to soft-resize it again in the editor, thinking the best behavior would be to do a visual "relative" resizing as that is WYSIWYG. Resizing an image and "guessing" how it would look is a pretty bad experience imho :)

An exception there is the direct size input. We should respect the values entered there in all cases. This also covers all cases where the user has uploaded an edited/resized image and intents to soft-resize it to a particular width.

To make the relative resizing of images that are narrower than the block width feel "proper", I think we should add a title to the 25-50-75-100% buttons, perhaps "Relative size", and disable the buttons that don't apply to the current image. I'll patch that so we can test how the UI "feels".

mor10 commented:
Correct me if I'm wrong here @azaozz: None of this has been merged into 5.0 RC1?

azaozz committed:
Update the 25-50-75-100% buttons

- Add a "Relative Image Size" label.
- DIsable buttons that don't apply when the image is narrower than the block width.

azaozz commented:
@mor10 correct. Ideally this will be merged together with a way for the theme to set uniform block_width for both the editor and the front-end (other than CSS).

markfinst commented:
What about adding `srcset` by default?

- Every image within Image or Gallery block has `srcset` and `sizes`.
- `srcset` contains list of all available image sizes which aspect ratio matches original. If user crops image - we return all images that match aspect ratio of cropped.
- `sizes` attribute is generated automatically depending on where image is added (for example, if image is added to full-width container sizes would be `100vw`).

To implement this, WordPress core badly needs an API method for `srcset` and `sizes`, specifically:

#### `get_image_srcset_attribute_array( $image_id, $max_image_width );`

Returns array of images that should be be used in srcset, in format:

```php
array(
  '123' => array(
    'url' => 'https://...',
    'width' => 600,
    'height' => 400,
    'size_key` => 'thumbnail'
  ),
 ...
)
```

So theme developers can easily filter it.

#### `get_image_sizes_attribute( $size_key, $size_ratio = 1  )`

- `$size_key` is a key that describes location of the image (`full-width`, `align-left`, `content` ...)
- `$size_ratio` is relative width multiplier, for example if image is added in 1/3 column, size_ratio should be `0.33`.

Should return `sizes` attribute value.

azaozz committed:
Check only for valid HTML 5.0 attribute names

All HTML attribute names match `/[a-z0-9-]+/` except `data-*`. Fix the check to do the same.

azaozz committed:
Update handling of "wide" and "full" alignment for the front-end

Update selecting a larger image when alignment is "wide" or "full". Look for larger sizes that may have been added by themes or plugins, then fall back to the original image (if larger). Keep the arbitrary limit of 4300px to avoid forcing site visitors to download huge image files.

azaozz committed:
Merge master

azaozz commented on packages/editor/src/utils/dom.js:
```diff
@@ -25,6 +25,39 @@ export function getBlockFocusableWrapper( clientId ) {
+
+			const measure = document.createElement( 'div' );
+			measure.className = 'wp-block editor-block-list__block';
+			layout.appendChild( measure );
```
To avoid the arbitrary `clientWidth - 30`, perhaps can append another div here with `class="editor-block-list__block-edit"` to mimic the default block layout, then get `clientWidth` from it. I'll try it.

I know, it's fragile, but seems workable for now.

azaozz committed:
Update `const getBlockWidth` to get the width from `div.editor-block-list__block-edit`

azaozz committed:
Fix typo in https://github.com/WordPress/gutenberg/pull/11973/commits/44c7c68ae058792d65e88553f526df44358d5b32

azaozz committed:
Fix missing space to make wpcs happy

aduth commented on lib/compat.php:
```diff
@@ -304,6 +304,286 @@ function gutenberg_warn_classic_about_blocks() {
+ */
+function gutenberg_render_block_core_image( $html, $block ) {
+	// Old post or... something's wrong.
+	if ( empty( $html ) || empty( $block['attrs'] ) ) {
```
I think we should be shortcutting here if the block is not the image block? It'll eventually be shortcutted at the below test of `empty( $block_attributes['url'] )`, but it seems both indirect and wasteful to let us get that far, when this is called for every block.

azaozz replied:
Right, good catch. Added in https://github.com/WordPress/gutenberg/pull/11973/commits/648811fd8c1bd3c62583c107928fada04d6af87e.

azaozz committed:
Update `gutenberg_render_block_core_image()`

Return early when the block is not `core/image`.

azaozz commented:
Heh, yoda conditionals fail for negative checks?? Who added that to the coding standards?? It was not a part of the original :)

Needs revision.

azaozz commented:
> One general concern I have here is scalability if we want to reproduce a similar behaviour to WordPress/twentynineteen#629 for other blocks that are not wp:image, including third-party blocks that might be using media.

Right, the idea was to add that to the Image Block first, test it, tweak it, and then add support for all images in all blocks. It won't be necessary for all images to have all block props, would be enough to at least have `width` and `height`, or `fileWidth` and `fileHeight`. 

Enabling support for other blocks would be as easy as changing (the negative yoda noncompliant) `$block['blockName'] !== 'core/image' ` to include more blocks :)

getsource commented:
@youknowriad I'm working on triage for WP 5.0.3, on https://core.trac.wordpress.org/ticket/45407, and noticed that this PR was moved out of all GB milestones.

Is still the approach planned to get the information to core? If it is, should it be milestoned somewhere so that it gets tracked? If not, I'd appreciate being linked to any other relevant issues or PRs so that I can add them to the core ticket.

Thanks so much!

youknowriad commented:
Correct if I'm wrong but I think we kind of agreed that this PR is not the right approach to solve these issues. At least we should break it into more discrete items and discuss the path forward.

About the milestoning. PRs are milestoned according the corresponding plugin release. This didn't seem ready for me for 4.9 or 5.0. But the priorities... are handled more precisely in issues.

At the moment we have two milestones:

 - 5.0.x for the most pressing things
 - 5.x for the ones after

We also have "Phase 2" for things that can come more down the road. I'd like us to work and triage and clarify these milestones more. In my personal priorities for phase2, I consider this meta issue as a good one to follow https://github.com/WordPress/gutenberg/issues/13113

mor10 commented:
From my perspective, this PR is indeed too wide in scope and much of what it tries to solve deserves to be broken into smaller pieces or reallocated to a wider conversation about media in general.

Specifically regarding https://core.trac.wordpress.org/ticket/45407, this is still a pressing issue that needs a resolution immediately. Afaik it is possible to resolve that specific issue as proposed in the Trac ticket without altering large parts of core or GB. The main issue in the past seemed to be it requires both changes to Core and GB, but since GB is now core that shouldn't be a blocker.

I strongly recommend flagging the RICG ticket as high priority as core is still shipping incorrect code and preventing theme and plugin developers from resolving it on their end.

gziolo commented:
This didn't get to WordPress 5.0 and 5.1. What are the next steps planned, should it be refreshed? @youknowriad and @mapk any thoughts? It also sort of duplicates #11377 as it was extracted from it.

aduth commented:
Per the most recent comments, it seems clear this pull request cannot be salvaged in its current state and will need to be revisited in separate smaller tasks.

Related tracking issues / tickets:

- #6177
- https://core.trac.wordpress.org/ticket/45407
